### ISSUE TITLE
Feature request: "np.permute"

### ISSUE THREAD

**Issue by mruberry** [numpy/numpy#17085]
Today in NumPy there's `transpose`, which "reverses or permutes" an array's axes. A "transposition," however, is typically a swap of two elements, like what `swapaxes` does. This issue proposes a new function, `permute`, which is equivalent to `transpose` except it requires the permutation be specified. There are two reasons for this proposal:

- It is the correct mathematical name for the operation
- It would be helpful to provide library writers a mechanism to permute both NumPy-like arrays and PyTorch tensors. PyTorch uses `transpose` for transpositions and `permute` for permutations. It plans to implement `swapaxes` as an alternative transposition mechanism, so `swapaxes` and `permute` would work on both PyTorch tensors and NumPy-like arrays (and make PyTorch tensors more NumPy-like). 

When considering new names I think natural questions are:

- Would a user expect a function with this name to do what it does?
- Does this name conflict with any existing functions?

In this case I think `permute`, as proposed, would do exactly what a user expects:  return the specified permutation of the array. The name also does not conflict with any names in NumPy or SciPy. There is `numpy.random.permutation`, which returns a permutation, but that seems natural in context:

```
np.permute(a, numpy.random.permutation(a.ndim))
```

would randomly permute the axes of an array.

I could implement this if the proposal is accepted.

**Comment by eric-wieser** [comment]
> I could implement this if the proposal is accepted.

It sounds like you're proposing
```python
def permute(a, perm):
    """Identical to transpose but the second argument is not optional"""
    return np.transpose(perm)
```
Am I understanding correctly?

**Comment by rkern** [comment]
I would find the name `permute()` to be confusing since we use similar language in the RNG to describe permutations of the _values_ of the array rather than the axes.

For what it's worth, there are usages in the mathematical literature of "transpose" to refer to any permutation of the axes of a tensor of arbitrary rank, so I don't think the first point is strictly true.

https://arxiv.org/abs/1411.1503
http://www.cs.cornell.edu/cv/SummerSchool/Contractions.pdf

**Comment by mruberry** [comment]
> > I could implement this if the proposal is accepted.
> 
> It sounds like you're proposing
> 
> ```python
> def permute(a, perm):
>     """Identical to transpose but the second argument is not optional"""
>     return np.transpose(a, perm)
> ```
> 
> Am I understanding correctly?

Yes.

**Comment by eric-wieser** [comment]
Agreed with @rkern that `permute_axis` would be better than `permute`.

But the question remains - who is helped by adding this new name? This feels very similar to #16469, but with a weaker justification.

**Comment by mruberry** [comment]
> Agreed with @rkern that `permute_axis` or `permute_dims` or `permute_axes` would be better than `permute`.
> 
> But the question remains - who is helped by adding this new name? This feels very similar to #16469, but with a weaker justification.

I think it'd mostly be for users familiar with programs like Keras, matlab and PyTorch and library writers looking to target multiple frameworks. Those three example framework are very popular. I think Keras is still the most popular deep learning framework and PyTorch is the most popular deep learning framework for research? I hear matlab's a thing, too.

In http://www.cs.cornell.edu/cv/SummerSchool/Contractions.pdf, by th eway, although the author starts by defining a "tensor transpose" he not only says "permutation" frequently but demonstrates his code with matlab's `permute` function. 

`permute_axes` would be fine for the library writer purpose and probably easy to discover for users familiar with other frameworks.

**Comment by rkern** [comment]
We usually resolve things like that with documentation pointing out the connections rather than adding additional aliases. NumPy is a mature package with a large API surface, and there is not a lot of appetite for expanding that API. We're trying to go in the other direction, in general. Adding a row to the table of [NumPy for MATLAB® Users](https://numpy.org/doc/stable/user/numpy-for-matlab-users.html) would be good.
